### PR TITLE
make replaceBookie() exclude ensemble bookie racks

### PR DESCRIPTION
**Pull request by sijie**
Descriptions of the changes in this PR:

This is a change ported back from yahoo/bookkeeper@0589d0d58d61364be1a897a5c1478d3109f4fc17. It excludes the ensemble bookie racks when replacing bookies. since there is already a new weight-based placement strategy introduced, so this change only applies when `weight-based` placement policy is not enabled.

This is a change contributed from Siddharth Boobna.

### CODE REVIEW

**Comment by sijie** [comment]
@merlimat can you eyeball this change to make sure I ported this correctly? since this change is conflicting with weight-based placement policy. so this feature is only *ON* when weight-based placement policy is disabled.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 1]
```diff
@@ -419,6 +418,14 @@ public void handleBookiesThatJoined(Set<BookieSocketAddress> joinedBookies) {
         return nodes;
     }
 
+    private Set<String> getNetworkLocations(Set<Node> bookieNodes) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java]: static ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 2]
```diff
@@ -594,8 +608,57 @@ public BookieNode selectFromNetworkLocation(
...
+                                                   Predicate<BookieNode> predicate,
+                                                   Ensemble<BookieNode> ensemble)
+            throws BKNotEnoughBookiesException {
+        // first attempt to select one from local rack
+        try {
+            return selectRandomFromRack(networkLoc, excludeBookies, predicate, ensemble);
+        } catch (BKNotEnoughBookiesException e) {
+            if (isWeighted) {
+                // if weight based selection is enabled, randomly select one from the whole cluster
+                // based on weights and ignore the provided predicate.
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java]: "ignore the provided predicate" -> maybe the comment is wrong ?  it seems to me that the predicate is applied in selectRandom

**Comment by sijie** [comment]
@eolivelli fixed your comments.

**Pull request closed without merging**